File: example-jv-03adapters/src/main/java/foo/bar/example/foreadapters/OG.java
package foo.bar.example.foreadapters;

import android.app.Application;

import java.util.HashMap;
import java.util.Map;

import co.early.fore.core.WorkMode;
import co.early.fore.core.logging.AndroidLogger;
import co.early.fore.core.time.SystemTimeWrapper;
import foo.bar.example.foreadapters.feature.playlist.PlaylistAdvancedModel;
import foo.bar.example.foreadapters.feature.playlist.PlaylistSimpleModel;

import static co.early.fore.core.Affirm.notNull;

/**
 *
 * OG - Object Graph, pure DI implementation
 *
 * Copyright © 2019 early.co. All rights reserved.
 */
public class OG {

    private static boolean initialized = false;
    private static final Map<Class<?>, Object> dependencies = new HashMap<>();


    public static void setApplication(Application application) {
        setApplication(application, WorkMode.ASYNCHRONOUS);
    }

    public static void setApplication(Application application, final WorkMode workMode) {

        notNull(application);
        notNull(workMode);


        // create dependency graph
        AndroidLogger logger = new AndroidLogger("fore_");
        SystemTimeWrapper systemTimeWrapper = new SystemTimeWrapper();
        PlaylistAdvancedModel playlistAdvancedModel = new PlaylistAdvancedModel(
                systemTimeWrapper,
                workMode,
                logger);
        PlaylistSimpleModel playlistSimpleModel = new PlaylistSimpleModel(
                workMode,
                logger);


        // add models to the dependencies map if you will need them later
        dependencies.put(SystemTimeWrapper.class, systemTimeWrapper);
        dependencies.put(PlaylistAdvancedModel.class, playlistAdvancedModel);
        dependencies.put(PlaylistSimpleModel.class, playlistSimpleModel);

    }

    public static void init() {
        if (!initialized) {
            initialized = true;

            // run any necessary initialization code once object graph has been created here

        }
    }

    public static <T> T get(Class<T> model) {

        notNull(model);
        T t = model.cast(dependencies.get(model));
        notNull(t);

        return t;
    }

    public static <T> void putMock(Class<T> clazz, T object) {

        notNull(clazz);
        notNull(object);

        dependencies.put(clazz, object);
    }

}

File: example-kt-03adapters/src/main/java/foo/bar/example/foreadapterskt/OG.kt
package foo.bar.example.foreadapterskt

import android.app.Application
import co.early.fore.core.WorkMode
import co.early.fore.kt.core.logging.AndroidLogger
import co.early.fore.kt.core.logging.Logger
import foo.bar.example.foreadapterskt.feature.playlist.mutable.MutablePlaylistModel
import foo.bar.example.foreadapterskt.feature.playlist.immutable.ImmutablePlaylistModel
import java.util.HashMap


/**
 * OG - Object Graph, pure DI implementation
 *
 * Copyright © 2015-2020 early.co. All rights reserved.
*/
@Suppress("UNUSED_PARAMETER")
object OG {

    private var initialized = false
    private val dependencies = HashMap<Class<*>, Any>()

    @JvmOverloads
    fun setApplication(application: Application, workMode: WorkMode = WorkMode.ASYNCHRONOUS) {

        // create dependency graph
        val logger = AndroidLogger("fore_")
        val playlistAdvancedModel = MutablePlaylistModel(
                logger
        )
        val playlistSimpleModel = ImmutablePlaylistModel(
                logger
        )


        // add models to the dependencies map if you will need them later
        dependencies[MutablePlaylistModel::class.java] = playlistAdvancedModel
        dependencies[ImmutablePlaylistModel::class.java] = playlistSimpleModel
        dependencies[Logger::class.java] = logger

    }


    fun init() {
        if (!initialized) {
            initialized = true

            // run any necessary initialization code once object graph has been created here

        }
    }


    /**
     * This is how dependencies get injected, typically an Activity/Fragment/View will call this
     * during the onCreate()/onCreateView()/onFinishInflate() method respectively for each of the
     * dependencies it needs.
     *
     * Can use a DI library for similar behaviour using annotations
     *
     * Will return mocks if they have been set previously in putMock()
     *
     *
     * Call it like this:
     *
     * <code>
     *     yourModel = OG[YourModel::class.java]
     * </code>
     *
     * If you want to more tightly scoped object, one way is to pass a factory class here and create
     * an instance where you need it
     *
     */
    @Suppress("UNCHECKED_CAST")
    operator fun <T> get(model: Class<T>): T = dependencies[model] as T

    fun <T> putMock(clazz: Class<T>, instance: T) {
        dependencies[clazz] = instance as Any
    }
}
